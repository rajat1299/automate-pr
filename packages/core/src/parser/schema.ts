import { z } from "zod";
import { PRAutomatorError } from "../error";

/**
 * File path validation regex
 * Allows:
 * - Alphanumeric characters
 * - Forward slashes for directories
 * - Hyphens and underscores
 * - File extensions
 * - No absolute paths or parent directory traversal
 */
const FILE_PATH_REGEX = /^(?!\/|.*\.\.\/)[a-zA-Z0-9\/\-_\.]+$/;

/**
 * Git diff validation regex
 * Ensures the diff follows basic Git diff format:
 * - Must start with diff --git
 * - Must contain @@ markers for chunk headers
 * - Must contain actual changes (+/- lines)
 */
const DIFF_REGEX = /^diff --git.*?(?:\n@@.*?@@.*?)+$/ms;

/**
 * Schema for individual file changes
 */
export const FileChangeSchema = z.object({
  path: z.string()
    .regex(FILE_PATH_REGEX, {
      message: "Invalid file path. Must be relative and contain only alphanumeric characters, slashes, hyphens, underscores, and dots."
    }),
  diff: z.string()
    .min(20, "Diff content too short")
    .regex(DIFF_REGEX, {
      message: "Invalid diff format. Must be a valid Git diff."
    }),
  oldPath: z.string()
    .regex(FILE_PATH_REGEX, {
      message: "Invalid old file path. Must be relative and contain only alphanumeric characters, slashes, hyphens, underscores, and dots."
    })
    .optional(),
  mode: z.enum(["create", "modify", "delete", "rename"])
    .default("modify"),
  content: z.string()
    .min(1, "File content cannot be empty")
    .optional(),
}).strict();

/**
 * Schema for PR metadata
 */
export const PRMetadataSchema = z.object({
  title: z.string()
    .min(10, "PR title too short")
    .max(80, "PR title too long")
    .regex(/^[a-zA-Z0-9\s\-_\.,:;!?'"()[\]{}]+$/, {
      message: "PR title contains invalid characters"
    }),
  description: z.string()
    .min(30, "PR description too short")
    .max(4000, "PR description too long"),
  branch: z.string()
    .regex(/^[a-zA-Z0-9\-_\/]+$/, {
      message: "Invalid branch name"
    })
    .optional(),
  labels: z.array(z.string())
    .optional(),
  assignees: z.array(z.string())
    .optional(),
  reviewers: z.array(z.string())
    .optional(),
}).strict();

/**
 * Main schema for AI response validation
 */
export const AIResponseSchema = z.object({
  files: z.array(FileChangeSchema)
    .min(1, "At least one file change is required")
    .max(20, "Too many file changes in a single PR"),
  metadata: PRMetadataSchema,
  reasoning: z.string()
    .min(50, "Reasoning too short")
    .max(2000, "Reasoning too long")
    .optional(),
}).strict();

/**
 * Type definitions for schema
 */
export type FileChange = z.infer<typeof FileChangeSchema>;
export type PRMetadata = z.infer<typeof PRMetadataSchema>;
export type AIResponse = z.infer<typeof AIResponseSchema>;

/**
 * Validate AI response with fallback parsing
 */
export function validateAIResponse(input: unknown): AIResponse {
  try {
    // First attempt: direct validation
    return AIResponseSchema.parse(input);
  } catch (error) {
    if (error instanceof z.ZodError) {
      // Second attempt: try to fix common issues
      const fixed = attemptResponseRecovery(input);
      try {
        return AIResponseSchema.parse(fixed);
      } catch (recoveryError) {
        throw new PRAutomatorError("Invalid AI response format", {
          cause: recoveryError,
          type: "VALIDATION_ERROR",
        });
      }
    }
    throw new PRAutomatorError("Failed to parse AI response", {
      cause: error,
      type: "PARSER_ERROR",
    });
  }
}

/**
 * Attempt to recover from common response format issues
 */
function attemptResponseRecovery(input: unknown): unknown {
  if (typeof input !== "object" || input === null) {
    throw new PRAutomatorError("Response must be an object", {
      type: "VALIDATION_ERROR",
    });
  }

  const result: Record<string, unknown> = {};

  // Handle files array
  if ("files" in input && Array.isArray((input as any).files)) {
    result.files = (input as any).files.map((file: any) => ({
      path: String(file.path || ""),
      diff: normalizeDiff(file.diff || ""),
      ...(file.oldPath && { oldPath: String(file.oldPath) }),
      ...(file.mode && { mode: String(file.mode) }),
      ...(file.content && { content: String(file.content) }),
    }));
  } else {
    result.files = [];
  }

  // Handle metadata
  if ("metadata" in input && typeof (input as any).metadata === "object") {
    const meta = (input as any).metadata;
    result.metadata = {
      title: String(meta.title || "").slice(0, 80),
      description: String(meta.description || ""),
      ...(meta.branch && { branch: String(meta.branch) }),
      ...(meta.labels && { labels: ensureStringArray(meta.labels) }),
      ...(meta.assignees && { assignees: ensureStringArray(meta.assignees) }),
      ...(meta.reviewers && { reviewers: ensureStringArray(meta.reviewers) }),
    };
  } else {
    throw new PRAutomatorError("Missing or invalid metadata", {
      type: "VALIDATION_ERROR",
    });
  }

  // Handle optional reasoning
  if ("reasoning" in input && (input as any).reasoning) {
    result.reasoning = String((input as any).reasoning);
  }

  return result;
}

/**
 * Normalize diff content to ensure it follows Git diff format
 */
function normalizeDiff(diff: string): string {
  // Remove any leading/trailing whitespace
  diff = diff.trim();

  // Ensure diff starts with "diff --git"
  if (!diff.startsWith("diff --git")) {
    diff = `diff --git a/file b/file\n${diff}`;
  }

  // Ensure diff has chunk headers
  if (!diff.includes("@@")) {
    diff = `${diff}\n@@ -1,1 +1,1 @@\n${diff}`;
  }

  return diff;
}

/**
 * Ensure value is an array of strings
 */
function ensureStringArray(value: unknown): string[] {
  if (Array.isArray(value)) {
    return value.map(String);
  }
  return [];
} 