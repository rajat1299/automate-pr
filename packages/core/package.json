{
  "name": "@automate-pr/core",
  "version": "0.1.0",
  "private": true,
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "scripts": {
    "build": "tsc",
    "dev": "tsc -w",
    "typecheck": "tsc --noEmit",
    "test": "vitest run",
    "test:watch": "vitest",
    "clean": "rm -rf dist"
  },
  "dependencies": {
    "dotenv": "^16.4.5",
    "zod": "^3.22.4"
  },
  "devDependencies": {
    "@types/node": "^20.11.24",
    "typescript": "^5.3.3",
    "vitest": "^3.0.4"
  }
} 