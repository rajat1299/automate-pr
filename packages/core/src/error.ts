export type ErrorType = "user" | "system" | "ai" | "github" | "validation";
export type ErrorSeverity = "info" | "warning" | "error" | "fatal";

export interface ErrorMetadata {
  code?: string;
  severity?: ErrorSeverity;
  source?: string;
  timestamp?: Date;
  [key: string]: unknown;
}

export class PRAutomatorError extends Error {
  public readonly timestamp: Date;

  constructor(
    public readonly type: ErrorType,
    message: string,
    public readonly metadata: ErrorMetadata = {}
  ) {
    super(message);
    this.name = "PRAutomatorError";
    this.timestamp = new Date();
    
    // Ensure stack trace points to the actual error location
    Error.captureStackTrace(this, this.constructor);
  }

  /**
   * Create a user error (invalid input, permissions, etc.)
   */
  static user(message: string, metadata: Omit<ErrorMetadata, "severity"> = {}) {
    return new PRAutomatorError("user", message, { ...metadata, severity: "error" });
  }

  /**
   * Create a system error (IO, network, etc.)
   */
  static system(message: string, metadata: Omit<ErrorMetadata, "severity"> = {}) {
    return new PRAutomatorError("system", message, { ...metadata, severity: "error" });
  }

  /**
   * Create an AI-related error (model, parsing, etc.)
   */
  static ai(message: string, metadata: Omit<ErrorMetadata, "severity"> = {}) {
    return new PRAutomatorError("ai", message, { ...metadata, severity: "error" });
  }

  /**
   * Create a GitHub API error
   */
  static github(message: string, metadata: Omit<ErrorMetadata, "severity"> = {}) {
    return new PRAutomatorError("github", message, { ...metadata, severity: "error" });
  }

  /**
   * Create a validation error (safety checks, linting, etc.)
   */
  static validation(message: string, metadata: Omit<ErrorMetadata, "severity"> = {}) {
    return new PRAutomatorError("validation", message, { ...metadata, severity: "error" });
  }

  /**
   * Convert error to a structured object for logging/reporting
   */
  toJSON() {
    return {
      type: this.type,
      message: this.message,
      timestamp: this.timestamp.toISOString(),
      stack: this.stack,
      ...this.metadata
    };
  }

  /**
   * Format error for display to users
   */
  format(): string {
    const parts = [
      `${this.type.toUpperCase()} ERROR: ${this.message}`
    ];

    if (this.metadata.code) {
      parts.push(`Code: ${this.metadata.code}`);
    }

    if (this.metadata.source) {
      parts.push(`Source: ${this.metadata.source}`);
    }

    return parts.join("\n");
  }
} 