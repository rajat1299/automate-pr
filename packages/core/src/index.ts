// Error handling
export { PRAutomatorError } from "./error";
export type { ErrorType, ErrorMetadata, ErrorSeverity } from "./error";

// Configuration
export { loadConfig } from "./config";
export type { Config } from "./config";

// Security
export { CredentialVault } from "./security/vault";
export type { VaultOptions } from "./security/vault";

export { PRWorkflow, WorkflowError } from "./workflow";
export type { WorkflowOptions } from "./workflow"; 