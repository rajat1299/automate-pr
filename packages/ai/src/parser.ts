import { AIResponse, AIResponseSchema } from "./schema/response";
import { PRAutomatorError } from "@automate-pr/core/error";

export class ResponseParser {
  /**
   * Extract JSON from markdown-like response
   */
  private static extractJSON(response: string): string {
    const jsonMatch = response.match(/```(?:json)?\n([\s\S]*?)\n```/);
    if (!jsonMatch) {
      throw new PRAutomatorError(
        "ai",
        "No JSON found in response"
      );
    }
    return jsonMatch[1].trim();
  }

  /**
   * Parse and validate AI response
   */
  static parse(response: string): AIResponse {
    try {
      // Extract JSON from response
      const jsonStr = this.extractJSON(response);

      // Parse JSON
      let data: unknown;
      try {
        data = JSON.parse(jsonStr);
      } catch (error) {
        throw new PRAutomatorError(
          "ai",
          "Failed to parse JSON from AI response",
          { error }
        );
      }

      // Validate against schema
      try {
        return AIResponseSchema.parse(data);
      } catch (error) {
        throw new PRAutomatorError(
          "ai",
          "Invalid AI response format",
          { error }
        );
      }
    } catch (error) {
      if (error instanceof PRAutomatorError) {
        throw error;
      }
      throw new PRAutomatorError(
        "ai",
        `Failed to parse AI response: ${error.message}`,
        { error }
      );
    }
  }

  /**
   * Parse unified diff format into structured changes
   */
  static parseDiff(diff: string): {
    chunks: Array<{
      oldStart: number;
      oldLines: number;
      newStart: number;
      newLines: number;
      content: string;
    }>;
    files: string[];
    additions: number;
    deletions: number;
  } {
    const chunks: Array<{
      oldStart: number;
      oldLines: number;
      newStart: number;
      newLines: number;
      content: string;
    }> = [];

    const files = new Set<string>();
    let additions = 0;
    let deletions = 0;

    try {
      // Split diff into chunks
      const diffChunks = diff.split(/(?=^@@)/m);

      for (const chunk of diffChunks) {
        // Skip empty chunks
        if (!chunk.trim()) continue;

        // Parse file names
        if (chunk.startsWith("---")) {
          const fileMatch = chunk.match(/^--- a\/(.*)/m);
          if (fileMatch) {
            files.add(fileMatch[1]);
          }
          continue;
        }

        // Parse chunk header
        const headerMatch = chunk.match(
          /^@@ -(\d+),?(\d+)? \+(\d+),?(\d+)? @@/
        );
        if (!headerMatch) continue;

        const [, oldStart, oldLines, newStart, newLines] = headerMatch;

        // Count changes
        const lines = chunk.split("\n").slice(1);
        for (const line of lines) {
          if (line.startsWith("+")) additions++;
          if (line.startsWith("-")) deletions++;
        }

        chunks.push({
          oldStart: parseInt(oldStart, 10),
          oldLines: oldLines ? parseInt(oldLines, 10) : 1,
          newStart: parseInt(newStart, 10),
          newLines: newLines ? parseInt(newLines, 10) : 1,
          content: chunk
        });
      }

      return {
        chunks,
        files: Array.from(files),
        additions,
        deletions
      };
    } catch (error) {
      throw new PRAutomatorError(
        "ai",
        `Failed to parse diff: ${error.message}`,
        { error, diff }
      );
    }
  }

  /**
   * Format PR title into conventional commit format
   */
  static formatCommitMessage(
    type: string,
    scope: string | undefined,
    breaking: boolean,
    title: string
  ): string {
    const parts = [type];
    
    if (scope) {
      parts.push(`(${scope})`);
    }
    
    if (breaking) {
      parts.push("!");
    }
    
    parts.push(": ", title);
    
    return parts.join("");
  }

  /**
   * Validate file changes against repository structure
   */
  static validateChanges(
    changes: AIResponse["files"],
    repoFiles: Set<string>
  ): {
    valid: boolean;
    errors: string[];
  } {
    const errors: string[] = [];

    for (const change of changes) {
      switch (change.action) {
        case "create":
          if (repoFiles.has(change.path)) {
            errors.push(
              `File ${change.path} already exists but marked for creation`
            );
          }
          if (!change.content) {
            errors.push(
              `No content provided for new file ${change.path}`
            );
          }
          break;

        case "modify":
        case "delete":
          if (!repoFiles.has(change.path)) {
            errors.push(
              `File ${change.path} does not exist but marked for ${change.action}`
            );
          }
          if (change.action === "modify" && !change.diff) {
            errors.push(
              `No diff provided for modified file ${change.path}`
            );
          }
          break;

        case "rename":
          if (!repoFiles.has(change.path)) {
            errors.push(
              `File ${change.path} does not exist but marked for rename`
            );
          }
          if (!change.oldPath) {
            errors.push(
              `No old path provided for renamed file ${change.path}`
            );
          }
          break;
      }
    }

    return {
      valid: errors.length === 0,
      errors
    };
  }
} 