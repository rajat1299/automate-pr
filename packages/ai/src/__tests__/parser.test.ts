import { describe, it, expect } from "vitest";
import { ResponseParser } from "../parser";
import { PRAutomatorError } from "@automate-pr/core/error";

describe("ResponseParser", () => {
  describe("parse", () => {
    it("parses valid AI response", () => {
      const response = `Here's the plan for your changes:

\`\`\`json
{
  "files": [
    {
      "path": "src/components/Button.tsx",
      "action": "modify",
      "diff": "@@ -1,4 +1,4 @@\\n-const Button = () => {\\n+const Button = ({ children }) => {\\n"
    }
  ],
  "pr": {
    "title": "fix: improve Button component props",
    "body": "Add support for children prop in Button component",
    "type": "fix",
    "scope": "components",
    "breaking": false
  },
  "reviewers": [
    {
      "username": "ui-team-lead",
      "reason": "UI component expertise",
      "expertise": ["react", "typescript", "design-systems"]
    }
  ],
  "metadata": {
    "confidence": 0.95,
    "reasoning": "Simple prop addition with no breaking changes",
    "suggestedLabels": ["components", "enhancement"],
    "estimatedComplexity": "low"
  }
}
\`\`\``;

      const result = ResponseParser.parse(response);
      expect(result).toMatchSnapshot();
    });

    it("handles missing JSON block", () => {
      const response = "Here's what we should do...";
      
      expect(() => ResponseParser.parse(response)).toThrow(
        new PRAutomatorError("ai", "No JSON found in response")
      );
    });

    it("handles invalid JSON", () => {
      const response = "\`\`\`json\ninvalid json\n\`\`\`";
      
      expect(() => ResponseParser.parse(response)).toThrow(
        new PRAutomatorError("ai", "Failed to parse JSON from AI response")
      );
    });

    it("validates against schema", () => {
      const response = `\`\`\`json
{
  "files": [],
  "pr": {
    "title": "x".repeat(100),
    "body": "test",
    "type": "invalid-type"
  }
}
\`\`\``;
      
      expect(() => ResponseParser.parse(response)).toThrow(
        new PRAutomatorError("ai", "Invalid AI response format")
      );
    });
  });

  describe("parseDiff", () => {
    it("parses unified diff format", () => {
      const diff = `--- a/src/components/Button.tsx
+++ b/src/components/Button.tsx
@@ -1,4 +1,4 @@
-const Button = () => {
+const Button = ({ children }) => {
   return (
-    <button>Click me</button>
+    <button>{children}</button>
   );
};`;

      const result = ResponseParser.parseDiff(diff);
      expect(result).toMatchSnapshot();
      expect(result.additions).toBe(2);
      expect(result.deletions).toBe(2);
      expect(result.files).toContain("src/components/Button.tsx");
    });

    it("handles multiple chunks", () => {
      const diff = `--- a/file.ts
+++ b/file.ts
@@ -1,2 +1,2 @@
-line1
+newline1
@@ -10,2 +10,2 @@
-line2
+newline2`;

      const result = ResponseParser.parseDiff(diff);
      expect(result.chunks).toHaveLength(2);
      expect(result.additions).toBe(2);
      expect(result.deletions).toBe(2);
    });

    it("handles invalid diff format", () => {
      const diff = "invalid diff";
      
      expect(() => ResponseParser.parseDiff(diff)).toThrow(
        new PRAutomatorError("ai", "Failed to parse diff")
      );
    });
  });

  describe("formatCommitMessage", () => {
    it.each([
      {
        type: "feat",
        scope: "ui",
        breaking: false,
        title: "add button component",
        expected: "feat(ui): add button component"
      },
      {
        type: "fix",
        scope: undefined,
        breaking: true,
        title: "change API response format",
        expected: "fix!: change API response format"
      },
      {
        type: "chore",
        scope: "deps",
        breaking: false,
        title: "update dependencies",
        expected: "chore(deps): update dependencies"
      }
    ])("formats $type commit with scope=$scope breaking=$breaking", ({
      type, scope, breaking, title, expected
    }) => {
      const result = ResponseParser.formatCommitMessage(type, scope, breaking, title);
      expect(result).toBe(expected);
    });
  });

  describe("validateChanges", () => {
    const repoFiles = new Set([
      "src/components/Button.tsx",
      "src/utils/index.ts"
    ]);

    it("validates valid changes", () => {
      const changes = [
        {
          path: "src/components/NewComponent.tsx",
          action: "create" as const,
          content: "// New component"
        },
        {
          path: "src/components/Button.tsx",
          action: "modify" as const,
          diff: "@@ -1 +1 @@\n-old\n+new"
        }
      ];

      const result = ResponseParser.validateChanges(changes, repoFiles);
      expect(result.valid).toBe(true);
      expect(result.errors).toHaveLength(0);
    });

    it("detects file existence conflicts", () => {
      const changes = [
        {
          path: "src/components/Button.tsx",
          action: "create" as const,
          content: "// Duplicate"
        },
        {
          path: "nonexistent.ts",
          action: "modify" as const,
          diff: "@@ -1 +1 @@\n-old\n+new"
        }
      ];

      const result = ResponseParser.validateChanges(changes, repoFiles);
      expect(result.valid).toBe(false);
      expect(result.errors).toHaveLength(2);
      expect(result.errors[0]).toContain("already exists");
      expect(result.errors[1]).toContain("does not exist");
    });

    it("validates required fields", () => {
      const changes = [
        {
          path: "new.ts",
          action: "create" as const
        },
        {
          path: "src/components/Button.tsx",
          action: "modify" as const
        },
        {
          path: "src/utils/index.ts",
          action: "rename" as const
        }
      ];

      const result = ResponseParser.validateChanges(changes, repoFiles);
      expect(result.valid).toBe(false);
      expect(result.errors).toHaveLength(3);
      expect(result.errors[0]).toContain("No content provided");
      expect(result.errors[1]).toContain("No diff provided");
      expect(result.errors[2]).toContain("No old path provided");
    });
  });
}); 