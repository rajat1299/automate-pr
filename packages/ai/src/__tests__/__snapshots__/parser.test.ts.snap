// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResponseParser > parse > parses valid AI response 1`] = `
{
  "files": [
    {
      "action": "modify",
      "diff": "@@ -1,4 +1,4 @@\\n-const Button = () => {\\n+const Button = ({ children }) => {\\n",
      "path": "src/components/Button.tsx",
    },
  ],
  "metadata": {
    "confidence": 0.95,
    "estimatedComplexity": "low",
    "reasoning": "Simple prop addition with no breaking changes",
    "suggestedLabels": [
      "components",
      "enhancement",
    ],
  },
  "pr": {
    "body": "Add support for children prop in Button component",
    "breaking": false,
    "scope": "components",
    "title": "fix: improve Button component props",
    "type": "fix",
  },
  "reviewers": [
    {
      "expertise": [
        "react",
        "typescript",
        "design-systems",
      ],
      "reason": "UI component expertise",
      "username": "ui-team-lead",
    },
  ],
}
`;

exports[`ResponseParser > parseDiff > parses unified diff format 1`] = `
{
  "additions": 2,
  "chunks": [
    {
      "content": "@@ -1,4 +1,4 @@
-const Button = () => {
+const Button = ({ children }) => {
   return (
-    <button>Click me</button>
+    <button>{children}</button>
   );
};",
      "newLines": 4,
      "newStart": 1,
      "oldLines": 4,
      "oldStart": 1,
    },
  ],
  "deletions": 2,
  "files": [
    "src/components/Button.tsx",
  ],
} 